### ISSUE TITLE
Google Maps overdraws a lot due to canvas context not being marked opaque

### ISSUE THREAD

**Issue by pcwalton** [servo/webrender#3018]
Maybe we should try to autodetect the case in which the canvas context is opaque to avoid overdraw.

**Comment by kvark** [comment]
Wouldn't this be Gecko responsibility to communicate to us that the canvas is opaque?

**Comment by pcwalton** [comment]
It's not clear to me how Gecko would do it right now. Also not clear to me how WR would do it…

**Comment by pcwalton** [comment]
Possibilities:

* Render content that we believe to be opaque (e.g. canvas) front-to-back, and use the Porter-Duff UNDER operator (with early Z enabled) for content beneath them. The downside is that we would conservatively have to turn on blending for any opaque objects behind such "probably-opaque" objects.

* Have the notion of a canvas background color, and promote canvases to opaque if they have a solid color behind them. (This is basically the old raytracing technique.)

**Comment by mstange** [comment]
> Render content that we believe to be opaque (e.g. canvas) front-to-back, and use the Porter-Duff UNDER operator (with early Z enabled) for content beneath them. The downside is that we would conservatively have to turn on blending for any opaque objects behind such "probably-opaque" objects.

This would also require us to do a clear-to-transparency at the start, and be opposed to the other clear-based optimization, wouldn't it?

**Comment by pcwalton** [comment]
Yes, it would.

**Comment by kvark** [comment]
@pcwalton 
> Render content that we believe to be opaque (e.g. canvas) front-to-back, and use the Porter-Duff UNDER operator (with early Z enabled) for content beneath them. 

We already render opaque stuff front-to-back, and we already use Z rejection (by opaque) for transparent things. So the only change from the current behavior is the UNDER operator?

**Comment by nical** [comment]
We can detect some of the opaque cases in gecko's canvas implementation the same way we try to detect when the first drawing command clobbers all pixels to skip the copy-on-write.
If possible I'd prefer to avoid complicated rendering tricks when compositing the canvas.

**Comment by jrmuizel** [comment]
Just to be clear, the situation with WebRender is no worse than what we have today right?

**Comment by nical** [comment]
> Just to be clear, the situation with WebRender is no worse than what we have today right?

To be overly specific, as far as I know canvases contribute the same amount of overdraw with and without WebRender. I don't know whether overdraw and performance in general are better or worse on google Maps if that's what you refer to as the situation.

**Comment by pcwalton** [comment]
Note that Google Maps is WebGL, so we need some way to detect when the
WebGL canvas is opaque...

**Comment by jrmuizel** [comment]
This comes through the 'alpha' option specified at context creation time: https://developer.mozilla.org/en-US/docs/Web/API/HTMLCanvasElement/getContext

**Comment by pcwalton** [comment]
Yeah, we do optimize based on that flag, but Google Maps unfortunately
doesn't set it.

**Comment by jrmuizel** [comment]
We should probably just ask them to set it then.

**Comment by pcwalton** [comment]
OK, I'll double check that they indeed aren't setting that flag and then
reach oy to my contacts.
